README.md
![](https://raw.githubusercontent.com/EFPrefix/EFMarkdown/master/Assets/EFMarkdown.png)

<p align="center">
    <a href="https://travis-ci.org/EFPrefix/EFMarkdown">
        <img src="http://img.shields.io/travis/EFPrefix/EFMarkdown.svg">
    </a>
    <a href="http://cocoapods.org/pods/EFMarkdown">
        <img src="https://img.shields.io/cocoapods/v/EFMarkdown.svg?style=flat">
    </a>
    <a href="http://cocoapods.org/pods/EFMarkdown">
        <img src="https://img.shields.io/cocoapods/p/EFMarkdown.svg?style=flat">
    </a>
    <a href="https://raw.githubusercontent.com/EFPrefix/EFMarkdown/master/LICENSE">
        <img src="https://img.shields.io/cocoapods/l/EFMarkdown.svg?style=flat">
    </a>
    <a href="https://twitter.com/EyreFree777">
        <img src="https://img.shields.io/badge/twitter-@EyreFree777-blue.svg?style=flat">
    </a>
    <a href="http://weibo.com/eyrefree777">
        <img src="https://img.shields.io/badge/weibo-@EyreFree-red.svg?style=flat">
    </a>
    <img src="https://img.shields.io/badge/made%20with-%3C3-orange.svg">
</p>

EFMarkdown is a lightweight Markdown library in Swift, based on [EFCMark](https://github.com/EFPrefix/EFCMark), inspired by [markdown](https://github.com/vapor-community/markdown) and [Markoff](https://github.com/thoughtbot/Markoff).

> [中文介绍](https://github.com/EFPrefix/EFMarkdown/blob/master/README_CN.md)

## Overview

sample1|sample2|sample3|sample4  
:---------------------:|:---------------------:|:---------------------:|:---------------------:
![](https://raw.githubusercontent.com/EFPrefix/EFMarkdown/master/Assets/sample1.png)|![](https://raw.githubusercontent.com/EFPrefix/EFMarkdown/master/Assets/sample2.jpg)|![](https://raw.githubusercontent.com/EFPrefix/EFMarkdown/master/Assets/sample3.png)|![](https://raw.githubusercontent.com/EFPrefix/EFMarkdown/master/Assets/sample4.jpg)  

## Example

To run the example project, clone the repo, and run `pod install` from the Example directory first.

## Requirements

| Version | Needs                                                         |
|:---------|:------------------------------------------------|
| 0.x        | XCode 8.0+<br>Swift 3.0+<br>iOS 8.0+   |
| 4.x        | XCode 9.0+<br>Swift 4.0+<br>iOS 8.0+   |
| 5.x        | XCode 10.2+<br>Swift 5.0+<br>iOS 8.0+ |

## Installation

EFMarkdown is available through [CocoaPods](http://cocoapods.org). To install
it, simply add the following line to your Podfile:

```ruby
pod "EFMarkdown"
```

## Usage

### 1. Markdown to HTML

You can use `EFMarkdown` to make Markdown string to HTML string easily:

```swift
let markdown = "# Hello"
var html = ""
do {
    html = try EFMarkdown().markdownToHTML(markdown, options: EFMarkdownOptions.safe)
    print(html) // This will return "<h1>Hello</h1>\n"
} catch let error as NSError {
    print ("Error: \(error.domain)")
}
```

### 2. View Markdown

You can use `EFMarkdownView` to make a preview of Markdown:

```swift
let screenSize = UIScreen.main.bounds
let markView = EFMarkdownView()
markView.frame = CGRect(x: 0, y: 20, width: screenSize.width, height: screenSize.height - 20)
markView.onRendered = {
    [weak self] (height) in
    if let _ = self {
        // Optional: you can know the change of height in this block
        print("onRendered height: \(height ?? 0)")
    }
}
self.view.addSubview(markView)
markView.load(markdown: testMarkdownFileContent(), options: [.default]) {
    [weak self] (_, _) in
    if let _ = self {
        // Optional: you can change font-size with a value of percent here
        markView.setFontSize(percent: 128)
        printLog("load finish!")
    }
}
```

### 3. Options

You can pass different options to the underlying `cmark` library. By default `safe` is passed.

The available options are:

* default
* sourcePos
* hardBreaks
* safe
* noBreaks
* validateUTF8
* smart
* githubPreLang
* liberalHtmlTag

For more information on the available options, see [`cmark`](https://github.com/github/cmark).

## Author

EyreFree, eyrefree@eyrefree.org

## License

<img src='http://www.wtfpl.net/wp-content/uploads/2012/12/logo-220x160.png' width='110' height='80'/>

EFMarkdown is available under the WTFPL license. See the LICENSE file for more info.


CHANGELOG.md
# Change Log

-----

## [5.0.0](https://github.com/EFPrefix/EFMarkdown/releases/tag/5.0.0) (2019-03-31)

* Upgrade to Swift 5.0.

---

## [4.1.0](https://github.com/EFPrefix/EFMarkdown/releases/tag/4.1.0) (2018-11-06)

#### Add

* Upgrade to Swift 4.2.

---

## [4.0.0](https://github.com/EFPrefix/EFMarkdown/releases/tag/4.0.0) (2017-09-30)

#### Fix

* Update to Swift 4.

---

## [0.0.5](https://github.com/EFPrefix/EFMarkdown/releases/tag/0.0.5) (2017-09-05)

#### Fix

* Fix issue [#2](https://github.com/EFPrefix/EFMarkdown/issues/21).

---

## [0.0.4](https://github.com/EFPrefix/EFMarkdown/releases/tag/0.0.4) (2017-08-26)

#### Fix

* Change license.

---

## [0.0.3](https://github.com/EFPrefix/EFMarkdown/releases/tag/0.0.3) (2017-08-25)

#### Add

* Add Markdown table support.

---

## [0.0.2](https://github.com/EFPrefix/EFMarkdown/releases/tag/0.0.2) (2017-08-19)

#### Add

* Add EFMarkdownView.

#### Fix

* Improve documentation.

---

## [0.0.1](https://github.com/EFPrefix/EFMarkdown/releases/tag/0.0.1) (2017-08-18)

First public release.


Example/Tests/Tests.swift
import UIKit
import XCTest
import EFMarkdown

class Tests: XCTestCase {
    
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    func testExample() {
        // This is an example of a functional test case.
        XCTAssert(true, "Pass")
    }
    
    func testPerformanceExample() {
        // This is an example of a performance test case.
        self.measure() {
            // Put the code you want to measure the time of here.
        }
    }
    
}


EFMarkdown/Classes/String+.swift
//
//  String+.swift
//  EyreFree
//
//  Created by EyreFree on 2017/8/23.
//
//  Copyright (c) 2017 EyreFree <eyrefree@eyrefree.org>
//
//  Everyone is permitted to copy and distribute verbatim or modified
//  copies of this license document, and changing it is allowed as long
//  as the name is changed.
//
//             DO WHAT THE FUCK YOU WANT TO PUBLIC LICENSE
//    TERMS AND CONDITIONS FOR COPYING, DISTRIBUTION AND MODIFICATION
//
//   0. You just DO WHAT THE FUCK YOU WANT TO.

import UIKit

extension String {

    // [] 操作符重载
    subscript(index: Int) -> Character {
        return self[self.index(self.startIndex, offsetBy: index)]
    }

    // 字符数量
    func count() -> Int {
        return self.count
    }

    // 是否符合输入的正则表达式
    func conform(regex: String) -> Bool {
        return NSPredicate(format: "SELF MATCHES %@", regex).evaluate(with: self)
    }

    // 子串数量
    func occurrencesOf(subString: String) -> Int {
        return self.components(separatedBy: subString).count - 1
    }

    // 替换某个子字符串为另一字符串
    func replace(_ string: String, with: String, options: CompareOptions? = nil) -> String {
        if let options = options {
            return self.replacingOccurrences(of: string, with: with, options: options, range: nil)
        }
        return self.replacingOccurrences(of: string, with: with)
    }

    // 替换前缀
    func replacePrefix(string: String, with: String) -> String {
        if self.hasPrefix(string) {
            return with + String(self.dropFirst(string.count()))
        }
        return self
    }

    // 替换尾缀
    func replaceSuffix(string: String, with: String) -> String {
        if self.hasSuffix(string) {
            return String(self.dropLast(string.count())) + with
        }
        return self
    }

    // 移除某个子串
    func remove(string: String) -> String {
        return self.replace(string, with: "")
    }

    // 移除某个前缀
    func removePrefix(string: String) -> String {
        return self.replacePrefix(string: string, with: "")
    }

    // 移除某个尾缀
    func removeSuffix(string: String) -> String {
        return self.replaceSuffix(string: string, with: "")
    }

    // 将多个连续重复字符变为一个
    func toOne() -> String {
        var outString = self
        let length = self.count
        for index in (1 ..< length).reversed() {
            if outString[index] == outString[index - 1] {
                outString.remove(at: outString.index(outString.startIndex, offsetBy: index))
            }
        }
        return outString
    }
}


EFMarkdown/Classes/EFMarkdown.swift
//
//  EFMarkdown.swift
//  EyreFree
//
//  Created by EyreFree on 2017/4/9.
//
//  Copyright (c) 2017 EyreFree <eyrefree@eyrefree.org>
//
//  Everyone is permitted to copy and distribute verbatim or modified
//  copies of this license document, and changing it is allowed as long
//  as the name is changed.
//
//             DO WHAT THE FUCK YOU WANT TO PUBLIC LICENSE
//    TERMS AND CONDITIONS FOR COPYING, DISTRIBUTION AND MODIFICATION
//
//   0. You just DO WHAT THE FUCK YOU WANT TO.

import Foundation
import EFCMark

public enum EFMarkdownError: Error {
    case conversionFailed
    case resourceFailed
}

public struct EFMarkdownOptions: OptionSet {
    public let rawValue: Int32

    public init(rawValue: Int32) {
        self.rawValue = rawValue
    }

    static public let `default` = EFMarkdownOptions(rawValue: 0)
    static public let sourcePos = EFMarkdownOptions(rawValue: 1 << 1)
    static public let hardBreaks = EFMarkdownOptions(rawValue: 1 << 2)
    static public let safe = EFMarkdownOptions(rawValue: 1 << 3)
    static public let noBreaks = EFMarkdownOptions(rawValue: 1 << 4)
    static public let validateUTF8 = EFMarkdownOptions(rawValue: 1 << 9)
    static public let smart = EFMarkdownOptions(rawValue: 1 << 10)
    static public let githubPreLang = EFMarkdownOptions(rawValue: 1 << 11)
    static public let liberalHtmlTag = EFMarkdownOptions(rawValue: 1 << 12)
}

@objc public class EFMarkdown: NSObject {
    
    public func markdownToHTML(_ markdown: String, options: EFMarkdownOptions = [.safe]) throws -> String {
        var buffer: String?
        try formatTable(markdown: markdown).withCString {
            guard let buf = cmark_markdown_to_html($0, Int(strlen($0)), options.rawValue) else {
                throw EFMarkdownError.conversionFailed
            }
            buffer = String(cString: buf)
            free(buf)
        }
        if let buffer = buffer {
            return buffer
        }
        throw EFMarkdownError.conversionFailed
    }
}


Example/EFMarkdown/AppDelegate.swift
//
//  AppDelegate.swift
//  EFMarkdown
//
//  Created by EyreFree on 08/18/2017.
//  Copyright (c) 2017 EyreFree. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?

    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {

        return true
    }

    func applicationWillResignActive(_ application: UIApplication) {

    }

    func applicationDidEnterBackground(_ application: UIApplication) {

    }

    func applicationWillEnterForeground(_ application: UIApplication) {

    }

    func applicationDidBecomeActive(_ application: UIApplication) {

    }

    func applicationWillTerminate(_ application: UIApplication) {

    }
}


EFMarkdown/Classes/EFMarkdownView.swift
//
//  EFMarkdownView.swift
//  EyreFree
//
//  Created by EyreFree on 2017/8/18.
//
//  Copyright (c) 2017 EyreFree <eyrefree@eyrefree.org>
//
//  Everyone is permitted to copy and distribute verbatim or modified
//  copies of this license document, and changing it is allowed as long
//  as the name is changed.
//
//             DO WHAT THE FUCK YOU WANT TO PUBLIC LICENSE
//    TERMS AND CONDITIONS FOR COPYING, DISTRIBUTION AND MODIFICATION
//
//   0. You just DO WHAT THE FUCK YOU WANT TO.

import UIKit
import WebKit

@objcMembers open class EFMarkdownView: UIView {

    // Main content view
    public let webView: WKWebView = WKWebView(
        frame: CGRect.zero, configuration: WKWebViewConfiguration()
    )

    // ScrollEnabled
    public var isScrollEnabled: Bool = true {
        didSet {
            webView.scrollView.isScrollEnabled = isScrollEnabled
        }
    }

    // Can link click jump, default allow all
    public var onTouchLink: ((URLRequest) -> Bool)?

    // New height callback, Default is nil
    public var onRendered: ((CGFloat?) -> Void)?

    // Load finish callback temp handler
    fileprivate var onFinishLoad: ((WKWebView, WKNavigation?) -> Void)?

    // Last height
    private var lastHeight: CGFloat = 0

    public convenience init() {
        self.init(frame: CGRect.zero)
    }

    override init (frame: CGRect) {
        super.init(frame : frame)
        setupViews()
    }

    public required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        setupViews()
    }

    func setupViews() {
        setupWebView()

        print("setupViews")

        addObserver(self, forKeyPath: "webView.scrollView.contentSize", options: .new, context: nil)
    }

    deinit {
        removeObserver(self, forKeyPath: "webView.scrollView.contentSize")
    }

    override open func observeValue(
        forKeyPath keyPath: String?,
        of object: Any?,
        change: [NSKeyValueChangeKey : Any]?,
        context: UnsafeMutableRawPointer?
        ) {
        if let change = change, let onRendered = onRendered, keyPath == "webView.scrollView.contentSize" {
            if let nsSize = change[NSKeyValueChangeKey.newKey] as? NSValue {
                if nsSize.cgSizeValue.height != lastHeight {
                    lastHeight = nsSize.cgSizeValue.height
                    onRendered(nsSize.cgSizeValue.height)
                }
            }
        }
    }

    func setupWebView() {
        self.webView.scrollView.isScrollEnabled = self.isScrollEnabled
        self.webView.translatesAutoresizingMaskIntoConstraints = false
        self.webView.navigationDelegate = self
        self.webView.frame = self.bounds
        addSubview(self.webView)

        // Fit to parent view
        self.webView.translatesAutoresizingMaskIntoConstraints = false
        let leftLc = NSLayoutConstraint(
            item: self.webView,
            attribute: .left,
            relatedBy: .equal,
            toItem: self,
            attribute: .left,
            multiplier: 1,
            constant: 0
        )
        let rightLc = NSLayoutConstraint(
            item: self.webView,
            attribute: .right,
            relatedBy: .equal,
            toItem: self,
            attribute: .right,
            multiplier: 1,
            constant: 0
        )
        let topLc = NSLayoutConstraint(
            item: self.webView,
            attribute: .top,
            relatedBy: .equal,
            toItem: self,
            attribute: .top,
            multiplier: 1,
            constant: 0
        )
        let bottomLc = NSLayoutConstraint(
            item: self.webView,
            attribute: .bottom,
            relatedBy: .equal,
            toItem: self,
            attribute: .bottom,
            multiplier: 1,
            constant: 0
        )
        self.addConstraint(leftLc)
        self.addConstraint(rightLc)
        self.addConstraint(topLc)
        self.addConstraint(bottomLc)

        self.webView.backgroundColor = self.backgroundColor
    }

    public func load(
        markdown: String?,
        options: EFMarkdownOptions = [.safe],
        completionHandler: ((WKWebView, WKNavigation?) -> Void)? = nil) {
        guard let markdown = markdown else {
            return
        }
        do {
            let pageContent = try markdownToHTMLPage(markdown, options: options)
            onFinishLoad = completionHandler
            webView.loadHTMLString(pageContent, baseURL: baseURL())
        } catch let error as NSError {
            printLog("Error: \(error.domain)")
        }
    }

    private func markdownToHTMLPage(_ markdown: String, options: EFMarkdownOptions = [.safe]) throws -> String {
        var htmlContent: String = ""
        do {
            try htmlContent = EFMarkdown().markdownToHTML(markdown, options: options)
        } catch {
            throw error
        }
        if let templateURL = baseURL() {
            let templateContent = try String(contentsOf: templateURL, encoding: String.Encoding.utf8)
            return templateContent.replacingOccurrences(of: "$PLACEHOLDER", with: htmlContent)
        }
        throw EFMarkdownError.conversionFailed
    }

    func baseURL() -> URL? {
        let bundle = Bundle(for: EFMarkdownView.self)
        if let templateURL = bundle.bundleIdentifier?.hasPrefix("org.cocoapods") == true
            ? bundle.url(forResource: "index", withExtension: "html", subdirectory: "EFMarkdown.bundle")
            : bundle.url(forResource: "index", withExtension: "html") {
            return templateURL
        }
        return nil
    }

    // Change font-size of text with scale
    public func setFontSize(percent: CGFloat, completionHandler: ((Any?, Error?) -> Void)? = nil) {
        let jsFontSize = "document.getElementsByTagName('body')[0].style.webkitTextSizeAdjust='\(percent)%'"
        self.webView.evaluateJavaScript(jsFontSize, completionHandler: { (result, error) in
            // Finish
            completionHandler?(result, error)

            // Refresh height
            if let onRendered = self.onRendered {
                onRendered(self.webView.scrollView.contentSize.height)
            }
        })
    }
}

extension EFMarkdownView: WKNavigationDelegate {

    public func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
        // Refresh height
        if let onRendered = self.onRendered {
            onRendered(self.webView.scrollView.contentSize.height)
        }

        // Load finish
        onFinishLoad?(webView, navigation)
        onFinishLoad = nil
    }

    public func webView(_ webView: WKWebView, decidePolicyFor navigationAction: WKNavigationAction, decisionHandler: @escaping (WKNavigationActionPolicy) -> Void) {
        switch navigationAction.navigationType {
        case .linkActivated:
            if let onTouchLink = onTouchLink {
                if onTouchLink(navigationAction.request) {
                    decisionHandler(.allow)
                } else {
                    decisionHandler(.cancel)
                }
            } else {
                decisionHandler(.allow)
            }
        default:
            decisionHandler(.allow)
        }
    }
}


EFMarkdown/Classes/EFMarkdown+Table.swift
//
//  EFMarkdown+Table.swift
//  EyreFree
//
//  Created by EyreFree on 2017/8/24.
//
//  Copyright (c) 2017 EyreFree <eyrefree@eyrefree.org>
//
//  Everyone is permitted to copy and distribute verbatim or modified
//  copies of this license document, and changing it is allowed as long
//  as the name is changed.
//
//             DO WHAT THE FUCK YOU WANT TO PUBLIC LICENSE
//    TERMS AND CONDITIONS FOR COPYING, DISTRIBUTION AND MODIFICATION
//
//   0. You just DO WHAT THE FUCK YOU WANT TO.

import Foundation

public extension EFMarkdown {

    private enum LineType: UInt8 {
        case normal = 0
        case delimiter = 1
        case headOrText = 2
    }

    private enum AlignType: String {
        case none = ""
        case left = " style=\"text-align:left\""
        case center = " style=\"text-align:center\""
        case right = " style=\"text-align:right\""
    }

    func formatTable(markdown: String) -> String {
        if let pre = preTable(markdown: markdown) {
            return pre.0.joined(separator: "\n")
        }
        return markdown
    }

    private func preTable(markdown: String) -> ([String], [LineType])? {
        var lines = markdown.components(separatedBy: "\n")
        if lines.count < 2 {
            return nil
        }
        var marks = [LineType](repeating: .normal, count: lines.count)
        for (index, line) in lines.enumerated() {
            if isTableHeaderOrTextLine(inputLine: line) {
                marks[index] = .headOrText
                if isTableDelimiterLine(inputLine: line) {
                    if index > 0 {
                        if .headOrText == marks[index - 1] {
                            marks[index] = .delimiter
                        }
                    }
                }
            }
        }
        var noTable = true
        for mark in marks {
            if .delimiter == mark {
                noTable = false
                break
            }
        }
        if noTable {
            return nil
        }

        let length = marks.count
        var index = 0
        repeat {
            if .delimiter == marks[index] {
                if compareSuitable(lines[index - 1], lines[index]) {
                    let algins = formatTableDelimiterLine(inputLine: lines[index])
                    if .headOrText == marks[index - 1] {
                        lines[index - 1] = "<table><thead>" + formatTableHeaderOrTextLine(inputLine: lines[index - 1], aligns: algins, isHead: true)
                    }
                    let oldLine = lines[index]
                    lines[index] = "</thead><tbody>"

                    if index + 1 < length {
                        for offset in (index + 1) ..< length {
                            if .normal != marks[offset] && compareSuitable(lines[offset], oldLine) {
                                lines[offset] = formatTableHeaderOrTextLine(inputLine: lines[offset], aligns: algins)
                            } else {
                                lines[offset - 1] += "</tbody></table>"
                                index = offset
                                break
                            }
                            index = offset
                        }
                        continue
                    }
                }
            }
            index += 1
        } while(index < length)
        return (lines, marks)
    }

    private func compareSuitable(_ one: String, _ two: String) -> Bool {
        return one.replace("\\|", with: "").removePrefix(string: "|").removeSuffix(string: "|").occurrencesOf(subString: "|")
            == two.replace("\\|", with: "").removePrefix(string: "|").removeSuffix(string: "|").occurrencesOf(subString: "|")
    }

    private func isTableDelimiterLine(inputLine: String) -> Bool {
        var line = inputLine.replace("\\|", with: "")
        if line.contains("||") || !(line.contains("|") && line.contains("-")) {
            return false
        }
        line = line.trimmingCharacters(in: .whitespacesAndNewlines).replace("\t", with: " ").toOne()
        if !line.hasPrefix("|") {
            line = "|" + line
        }
        if !line.hasSuffix("|") {
            line = line + "|"
        }
        if line.conform(regex: "(\\| *:?-+:? *)+\\|") {
            return true
        }
        return false
    }

    private func isTableHeaderOrTextLine(inputLine: String) -> Bool {
        return inputLine.replace("\\|", with: "").contains("|")
    }

    private func formatTableDelimiterLine(inputLine: String) -> [AlignType] {
        var line = inputLine.trimmingCharacters(in: .whitespacesAndNewlines)
            .replace(" ", with: "").replace("\t", with: "").toOne()
        if !line.hasPrefix("|") {
            line = "|" + line
        }
        if !line.hasSuffix("|") {
            line = line + "|"
        }
        let lineArray = line.components(separatedBy: "|").dropFirst().dropLast()
        var alignArray = [AlignType]()
        for ele in lineArray {
            switch ele {
            case "-":
                alignArray.append(.none)
            case ":-":
                alignArray.append(.left)
            case "-:":
                alignArray.append(.right)
            case ":-:":
                alignArray.append(.center)
            default:
                alignArray.append(.none)
            }
        }
        return alignArray
    }

    private func formatTableHeaderOrTextLine(inputLine: String, aligns: [AlignType], isHead: Bool = false) -> String {
        let tempBlank = "/E+F!M@a#r$k%d^o&w-n*"
        var line = inputLine.replace("\\|", with: tempBlank).trimmingCharacters(in: .whitespacesAndNewlines)
        if !line.hasPrefix("|") {
            line = "|" + line
        }
        if !line.hasSuffix("|") {
            line = line + "|"
        }
        let lineArray = line.components(separatedBy: "|").dropFirst().dropLast()
        var lineHtml = "<tr>"
        let mark = isHead ? "th" : "td"
        for (index, ele) in lineArray.enumerated() {
            var eleHtml = ele
            do {
                eleHtml = try self.markdownToHTML(ele)
            } catch let error as NSError {
                print ("Error: \(error.domain)")
            }
            lineHtml += "<\(mark)\(aligns[index].rawValue)>\(eleHtml)</\(mark)>"
        }
        return (lineHtml + "</tr>").replace(tempBlank, with: "\\|")
    }
}


Example/EFMarkdown/ViewController.swift
//
//  ViewController.swift
//  EFMarkdown
//
//  Created by EyreFree on 08/18/2017.
//  Copyright (c) 2017 EyreFree. All rights reserved.
//

import UIKit
import EFMarkdown

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()

        // 1. EFMarkdown
        let markdown = "# Hello"
        var html = ""
        do {
            html = try EFMarkdown().markdownToHTML(markdown, options: EFMarkdownOptions.safe)
            // This will return "<h1>Hello</h1>\n"
            print(html)
        } catch let error as NSError {
            print ("Error: \(error.domain)")
        }

        // 2. EFMarkdownView
        let screenSize = UIScreen.main.bounds
        let markView = EFMarkdownView()
        markView.frame = CGRect(x: 0, y: 20, width: screenSize.width, height: screenSize.height - 20)
        markView.onRendered = {
            [weak self] (height) in
            if let _ = self {
                // Optional: you can know the change of height in this block
                print("onRendered height: \(height ?? 0)")
            }
        }
        self.view.addSubview(markView)
        markView.load(markdown: testMarkdownFileContent(), options: [.default]) {
            [weak self] (_, _) in
            if let _ = self {
                // Optional: you can change font-size with a value of percent here
                markView.setFontSize(percent: 128)
                printLog("load finish!")
            }
        }
    }

    public func testMarkdownFileContent() -> String {
        if let templateURL = Bundle.main.url(forResource: "sample1", withExtension: "md") {
            do {
                return try String(contentsOf: templateURL, encoding: String.Encoding.utf8)
            } catch {
                return ""
            }
        }
        return ""
    }
}
